ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt

```diff
@@ -233,27 +244,34 @@ data class PathSegmentParameterRouteSelector(val name: String, val prefix: Strin
  * @param suffix is an optional prefix
  */
 data class PathSegmentOptionalParameterRouteSelector(val name: String, val prefix: String? = null, val suffix: String? = null) : RouteSelector(RouteSelectorEvaluation.qualityParameter) {
+
     override fun evaluate(context: RoutingResolveContext, segmentIndex: Int): RouteSelectorEvaluation {
-        if (segmentIndex < context.segments.size) {
-            val part = context.segments[segmentIndex]
-            val prefixChecked = if (prefix == null)
-                part
-            else
-                if (part.startsWith(prefix))
-                    part.drop(prefix.length)
-                else
-                    return RouteSelectorEvaluation.Missing
-
-            val suffixChecked = if (suffix == null)
-                prefixChecked
-            else
-                if (prefixChecked.endsWith(suffix))
-                    prefixChecked.dropLast(suffix.length)
-                else
-                    return RouteSelectorEvaluation.Missing
+        return evaluate(context.segments, segmentIndex)
+    }
+
+    internal fun evaluate(segments: List<String>, segmentIndex: Int): RouteSelectorEvaluation {
```

Comment:
Is this code duplicate? Can extract the top-level method instead?

Reply:
Yes and no. In case of failure one method returns `RouteSelectorEvaluation.Failed` and another `RouteSelectorEvaluation.Missing`. They can be extracted to one with extra parameter, I will do it.